File: python/linked-lists/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        prev, curr, nxt = None, head, head.next

        while nxt:
            curr.next = prev
            prev = curr
            curr = nxt
            nxt = curr.next

        curr.next = prev

        delete = curr
        if n == 1:
            curr = curr.next
        else:
            for i in range(n - 2):
                delete = delete.next

            delete.next = delete.next.next

        head = curr
        if not head:
            return

        prev, curr, nxt = None, head, head.next

        while nxt:
            curr.next = prev
            prev = curr
            curr = nxt
            nxt = curr.next

        curr.next = prev

        return curr

File: c++/linked-list/remove-nth-node-from-end-of-list.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* reverse = reverseList(head);
        ListNode* pointer = reverse;

        if (n == 1) {
            return reverseList(reverse->next);
        }

        int count = 0;
        while (count < n - 2) {
            pointer = pointer->next;
            count += 1;
        }

        if (pointer->next) {
            pointer->next = pointer->next->next;
        } else {
            pointer->next = nullptr;
        }

        return reverseList(reverse);
    }

private:
    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next) {
            return head;
        }

        ListNode* prev = head;
        ListNode* curr = prev->next;

        while (curr) {
            ListNode* next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        head->next = nullptr;

        return prev;
    }
};
